068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,251 @@
+
+```java
+public class TopicAccess
+{
+  // topic name or pattern
+  public final String pattern;
+  // allowed operation
+  public final AclOperation operation;
+}
+
+public class JWTAuthorizer implements ClusterMetadataAuthorizer {
+    private StandardAuthorizer delegate;
+    private KeycloakRBACAuthorizer singleton;
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        // ...
+    }
+
+    @Override public List<AuthorizationResult> authorize(AuthorizableRequestContext requestContext, List<Action> actions) {
+        KafkaPrincipal principal = requestContext.principal();
+        if (!(principal instanceof OAuthKafkaPrincipal)) {
+          // simple ACL delegation
+        }
+        BearerTokenWithPayload token = ((OAuthKafkaPrincipal) principal).getJwt();
+        // check token validity
+        BearerTokenWithPayload jwt = principal.getJwt();
+        ObjectNode payload = jwt.getClaimsJSON();
+
+        // get the acls from the payload
+        List<TopicAccess> topicAccesses = extractTopicAccesses(topicAccessesRaw, prefix);
+        List<ClusterAccess> clusterAccesses = extractClusterAccesses(clusterAccessesRaw, prefix);
+
+        List<AuthorizationResult> results = new ArrayList<>(actions.size());
+        for (Action action : actions) {
+          if (checkTopicJwtClaims(topicAccesses, action) || checkClusterJwtClaims(clusterAccesses, action)) {
+              results.add(AuthorizationResult.ALLOWED);
+          } else {
+            results.add(AuthorizationResult.DENIED);
+          }
+        }
+        return results;
+    }
+
+    // sketch of how the topic access check could look like
+    public static boolean checkTopicJwtClaims(List<TopicAccess> topicAccesses, Action requestedAction) {
+    for (TopicAccess t : topicAccesses) {
+      switch (requestedAction.resourcePattern().resourceType()) {
+          case TOPIC:
+              if (matchTopicPattern(requestedAction, t) && checkTopicAccess(t.operation, requestedAction))
+                  return true;
+              break;
+          case CLUSTER: // ...
+          case GROUP: // ...
+      }
+      return false;
+    }
+
+    private static boolean checkTopicAccess(AclOperation claimedOperation, Action requestedAction) {
+      switch (requestedAction.operation()) {
+        case READ:
+          return List.of(ANY, ALL, READ).contains(claimedOperation);
+        case WRITE:
+        // ...
+      }
+    }
+
+    // ...
+}
+```
+
+Operations are described in the [Kafka docs](https://kafka.apache.org/documentation/#operations_in_kafka).
+Allowing an operation will follow the same logic as in [AclEntry.supportedOperations](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/security/authorizer/AclEntry.scala#L99).
+
+The above code will be supplied with unit tests and documentation explaining how to set up Strimzi with
+the new authorizer using Azure Active Directory B2C as an example.
```

Comment:
Can we also add a Keycloak example? Some users may want to switch to this new authorizer, and that would make their life easier, and increase adoption.

Reply:
That makes sense to me